Add more configuration options to wifi::ClientConfiguration

embediver opened the pull request:
Adds fields to configure Protected Management Frames (PMF) and scan method to `wifi::ClientConfiguration`.

PMF can be configured as
```rust
pub enum PmfConfiguration {
    /// No support for PMF will be advertized (default)
    #[default]
    NotCapable,
    /// Advertize PMF support and wether PMF is required or not
    Capable { required: bool },
}
```

and scan methods are also available the way esp-idf exposes them:
```rust
pub enum ScanMethod {
    /// Scan every channel and connect according to [ScanSortMethod]
    FullScan(ScanSortMethod),
    /// Connect to the first found AP and stop scanning (default)
    #[default]
    FastScan,
}

pub enum ScanSortMethod {
    /// Sort by signal strength
    Signal,
    /// Sort by Security (default)
    #[default]
    Security,
}
```
they are also marked `non_exhaustive` since other scan methods may be possible.

ivmarkov commented:
Looks quite clean already
Just one note: perhaps full scan should be the default, for backwards compatibility?

ivmarkov commented:
I guess we need to merge this after the corresponding changes in `esp-idf-svc` are also done. Agreed?

Vollbrecht commented:
we should probably check if we break stuff in esp-wifi so they have time to adapt

embediver commented:
Yes, we should get PRs in `esp-idf-svc` and `esp-wifi` ready before we merge this. 

I will also add some `strum` stuff to the Enums I created before we merge this.

embediver commented:
I added a bunch of `strum` derives and strum attributes to the Enums.
Renamed `FullScan` to `CompleteScan` as it is more appealing IMO.

But I'm not quite sure if I have the serialize strings all in form, which conforms to the rest of the crate.
Would be nice, if you could take a look at that.

Also it bothers me a bit that  the strings one adds with strum aren't automatically shown in the docs somewhere (wonder if that would be possible to implement in strum itself).

ivmarkov commented:
> I added a bunch of `strum` derives and strum attributes to the Enums. Renamed `FullScan` to `CompleteScan` as it is more appealing IMO.
> 
> But I'm not quite sure if I have the serialize strings all in form, which conforms to the rest of the crate. Would be nice, if you could take a look at that.

I think it looks good.

> Also it bothers me a bit that the strings one adds with strum aren't automatically shown in the docs somewhere (wonder if that would be possible to implement in strum itself).

I'm not sure what the future of the strum-related macros in this crate is. It has a very limited use in `edge-frame` that nobody besides me is using. So let's leave it at that.
